# PopGenCode
Various bits of code for population genetic methods used in my publications

## Compute D<sub>xy</sub> from site frequency spectrum

Code used to compute D<sub>xy</sub> in 10kb-windows across the genome in Marques et al. 2018 in press.

### 1) Estimate 2D-SFS in windows using angsd

I used the code in [run10kdxy.pbs](http://github.com/marqueda/PopGenCode/blob/master/run10kdxy.pbs) to compute 2D-SFS in predefined windows given in a BED-format file (e.g. file windows.chrI.GLA.10k.bed listing all 10kb-windows on chromosome I). For each line in the BED-file, the script runs angsd twice to generate a temporary SAF file for the specific window for both populations (defined by bam file lists) and then runs realSFS, which computes the 2D-SFS. Chromsome, start and end position of the window and the realSFS output are then added to the outfile, one line per line (see file [example.10kwin.sfs](https://github.com/marqueda/PopGenCode/blob/master/run10kdxy.pbs)). Sorry for the hard-coded format...

### 2) Compute D<sub>xy</sub> from 2D-SFS file

Once you have a file generated by the above script with chromosome, start, end position and SFS entries (file [example.10kwin.sfs](http://github.com/marqueda/PopGenCode/blob/master/run10kdxy.pbs)), the script [dxy_wsfs.py](http://github.com/marqueda/PopGenCode/blob/master/dxy_wsfs.py) computes D<sub>xy</sub> for each line in the file.
